docs/ds/queue.md

```diff
@@ -1,144 +1,144 @@
+Query front/end of queue: `q[ql]` / `q[qr]` 
 
-清空队列： `ql=1;qr=0;` 
+Empty deque: `ql=1;qr=0;` 
 
-数组模拟双端队列是同理的。
+Above rules work the same for array simulation deque.
 
-## 循环队列
+## Circular queue
 
-这样会导致一个问题：随着时间的推移，整个队列会向数组的尾部移动，一旦到达数组的最末端，即使数组的前端还有空闲位置，再进行入队操作也会导致溢出。（这种数组上实际有空闲位置而发生了上溢的现象称为是“假溢出”。
+This will cause a problem: as time goes by, the entire queue will move to the end of the array. Once it reaches the end, even if there is a empty place at the front of the array, the enqueue operation will cause overflow. (This kind of situation where there is actually empty space in the array and overflow still occurs is called "false overflow".)
 
-解决假溢出的办法是采用循环的方式来组织存放队列元素的数组，即将数组下标为 0 的位置看做是最后一个位置的后继。（ `x` 的后继为 `(x + 1) % Size` ）。这样就形成了循环队列。
+The solution to false overflow is to organize the array storing the queue elements in a circular manner, that is, the 0 index of the array is regarded as the successor of the last position. (The successor of `x` is `(x + 1)% Size` ). This would form a circular queue.
 
-## 双栈模拟队列
+## Double stack simulates queue
 
-其实不仅仅可以用数组模拟队列，还有一种冷门的方法是双栈模拟队列。
+In fact, not only arrays can be used to simulate queues, there is also a less popular method which is using double stacks to simulate queues.
 
-我们使用两个栈 F,S 模拟一个队列，其中 F 是队尾的栈，S 代表队首的栈，支持 push（在队尾插入），pop（在队首弹出）操作：
+We use two stacks F and S to simulate a queue, where F is the stack at the end of the queue, and S is the stack at the head of the queue. Push (insert at the end of the queue) and pop (pop at the beginning of the queue) operations are supported:
 
-1. Push：插入到栈 F 中
-2. Pop：如果 S 非空，让 S 弹栈；否则把 F 的元素倒过来圧到 S 中（其实就是一个一个弹出插入，做完后是首位颠倒的），然后再让 S 弹栈。
+1. Push: insert into stack F.
+2. Pop: if S is not empty, pop S; otherwise, turn the elements of F upside down and push them into S (which just means pop them one by one and insert, so that the first position is reversed after completion), and then pop S.
 
-容易证明，每个元素只会进入/转移/弹出一次，均摊复杂度 $O(1)$ 。
+It is easy to prove that each element will only enter/transfer/pop once, and the amortized time complexity is $O(1)$ .
 
-有人问这个东西有什么用吗？参见下面这道题。这道题顺便可以给大家一个 **双栈模拟双端队列** 的方法。
+Someone might be wondering why this is usaful? Please see the problem below. It can also help you understand the **double stacks simulate deque** method.
 
-## 例题
+## Sample problem
 
- [LOJ6515「雅礼集训 2018 Day10」贪玩蓝月](https://loj.ac/problem/6515) 
+ [LOJ6515 'Yali Training 2018 Day10' Playful Blue Moon](https://loj.ac/problem/6515) (Original link in Chinese)
 
-> 一个双端队列（deque），m 个事件：
+> One deque, m operations:
 >
-> 1. 在前端插入 (w,v)
-> 2. 在后端插入 (w,v)
-> 3. 删除前端的二元组
-> 4. 删除后端的二元组
-> 5.  给定 l,r，在当前 deque 中选择一个子集 S 使得 $\sum_{(w,v)\in S}w\bmod p\in[l,r]$ ，且最大化 $\sum_{(w,v)\in S}v$ .
+> 1.  Insert (w,v) at the front:
+> 2.  Insert (w,v) in the end;
+> 3.  Delete the tuple at the front;
+> 4.  Delete the tuple in the end;
+> 5.  For l, r, choose a subset S from deque so $\sum_{(w,v)\in S}w\bmod p\in[l,r]$ , and optimize $\sum_{(w,v)\in S}v$ .
```

Comment:
```suggestion
> 5.  Given l and r, choose a subset S from deque so that $\sum_{(w,v)\in S}w\bmod p\in[l,r]$ , and maximize $\sum_{(w,v)\in S}v$ .
```